components/board/client.go

```diff
@@ -245,14 +266,139 @@ func (dic *digitalInterruptClient) Tick(ctx context.Context, high bool, nanoseco
 	panic(errUnimplemented)
 }
 
-func (dic *digitalInterruptClient) AddCallback(c chan Tick) {
+func (dic *digitalInterruptClient) AddCallback(ch chan Tick) {
 	panic(errUnimplemented)
 }
 
-func (dic *digitalInterruptClient) RemoveCallback(c chan Tick) {
+func (dic *digitalInterruptClient) RemoveCallback(ch chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	stream := &interruptStream{
+		extra:  ext,
+		client: c,
+	}
+
+	err = stream.startStream(ctx, interrupts, ch)
+
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	c.interruptStreams = append(c.interruptStreams, stream)
+	c.mu.Unlock()
+
+	return nil
+
+}
+
+func (s *interruptStream) startStream(ctx context.Context, interrupts []string, ch chan Tick) error {
+	s.streamMu.Lock()
+	defer s.streamMu.Unlock()
+
+	s.streamRunning = true
+	s.streamReady = make(chan bool, 1024)
+	s.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	s.cancelBackgroundWorkers = cancel
+
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer s.activeBackgroundWorkers.Done()
+		s.connectTickStream(ctx, interrupts, ch)
+	})
+
+	//wait until the stream is ready to return
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.streamReady:
+		return nil
+
+	}
+}
+
+func (s *interruptStream) connectTickStream(ctx context.Context, interrupts []string, ch chan Tick) {
+	defer func() {
+		s.streamMu.Lock()
+		defer s.streamMu.Unlock()
+		s.streamCancel = nil
+		s.streamRunning = false
+	}()
+
+	streamCtx, cancel := context.WithCancel(ctx)
+	s.streamCancel = cancel
+
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+
+	req := &pb.StreamTicksRequest{
+		Name:     s.client.info.name,
+		PinNames: interrupts,
+	}
+
+	//call the server to start streaming ticks
+	stream, err := s.client.client.StreamTicks(streamCtx, req)
+	if err != nil {
+		s.client.logger.CError(ctx, err)
+		return
+	}
+
+	if s.streamReady != nil {
+		close(s.streamReady)
+	}
+	s.streamReady = nil
+
+	// repeatly recieve from the stream
+	for {
+		select {
+		case <-ctx.Done():
+			s.client.logger.CError(ctx, err)
+			s.closeStream()
+			return
+		default:
+		}
+		streamResp, err := stream.Recv()
+		if err != nil {
+			// only debug log the context canceled error
+			s.client.logger.Debug(err)
+			s.closeStream()
+			return
+		} else {
+			// If there is a response with a tick, send it to the channel
+			tick := Tick{
+				Name:             streamResp.PinName,
+				High:             streamResp.High,
+				TimestampNanosec: streamResp.Time,
+			}
+			ch <- tick
+		}
+	}
+}
+
+func (s *interruptStream) closeStream() {
+	s.streamCancel()
```

Comment:
Line 331 shows that this might become `nil`. I think this line can panic.

edit: no, `closeStream` can only be called _before_ line 331 runs. I don't get the control flow here; something can be simplified.